File: windows/RNSoundModule/RNSoundModule/RNSound.cs
using ReactNative.Bridge;
using ReactNative.Collections;
using System;
using System.Collections.Generic;
using Windows.ApplicationModel.Core;
using Windows.UI.Core;
using Windows.UI.Popups;
using Windows.Media.Capture;
using Windows.Media.MediaProperties;
using Windows.Storage;
using Windows.Storage.Streams;
using Windows.UI.Xaml.Controls;
using System.Threading.Tasks;
using System.Diagnostics;
using System.IO;
using ReactNative.Modules.Core;
using System.Threading;
using Newtonsoft.Json.Linq;
using Windows.UI.Xaml;
using System.Windows;
using Windows.UI.Xaml.Media;
using Windows.Media.Playback;
using Windows.Media.Core;
using System.Text;

namespace RNSoundModule
{
    public class RNSound : ReactContextNativeModuleBase
    {
        private const String IsWindows = "IsWindows";
        private ReactContext context;

        Dictionary<int, MediaPlayer> playerPool = new Dictionary<int, MediaPlayer>();
        static Object NULL = null;



        public RNSound(ReactContext reactContext)
            : base(reactContext)
        {
            context = reactContext;
        }
        public override string Name
        {
            get
            {
                return "RNSound";
            }
        }
        public override void Initialize()
        {
        }

        public override IReadOnlyDictionary<string, object> Constants
        {
            get
            {
                return new Dictionary<string, object>
                {
                    { IsWindows, true },
                };
            }
        }


        [ReactMethod]
        public async void prepare(String fileName, int key, JObject options, ICallback callback)
        {
            bool enableSMTCIntegration = true;
            JToken smtcOptionToken = options["enableSMTCIntegration"];
            if (smtcOptionToken != null)
            {
                enableSMTCIntegration = smtcOptionToken.Value<bool>();
            }

            await CoreApplication.MainView.CoreWindow.Dispatcher.RunAsync(CoreDispatcherPriority.Normal, async () =>
            {
                MediaPlayer player = await createMediaPlayer(fileName, enableSMTCIntegration);
                player.MediaOpened +=
                    delegate
                    {
                        JObject props = new JObject();
                        props.Add("duration", player.PlaybackSession.NaturalDuration.TotalMilliseconds * .001);
                        callback.Invoke(NULL, props);
                    };
                player.AutoPlay = false;
                if (player == null)
                {
                    JObject e = new JObject();

                    e.Add("code", -1);
                    e.Add("message", "resource not found");
                    callback.Invoke(e);
                    return;
                }

                this.playerPool.Add(key, player);
            });
        }

        protected async Task<MediaPlayer> createMediaPlayer(String fileName, bool enableSMTCIntegration)
        {
            MediaPlayer player = new MediaPlayer();
            player.CommandManager.IsEnabled = enableSMTCIntegration;

            StorageFile file = null;
            await CoreApplication.MainView.CoreWindow.Dispatcher.RunAsync(CoreDispatcherPriority.Normal,
                async () =>
                {
                    StorageFolder LocalFolder = ApplicationData.Current.LocalFolder;
                    StorageFolder InstallationFolder = Windows.ApplicationModel.Package.Current.InstalledLocation;

                    try
                    {
                        file = await InstallationFolder.GetFileAsync(@"Assets\" + fileName);
                    }
                    catch (Exception e) { }

                    if (file == null)
                    {
                        try
                        {
                            file = await LocalFolder.GetFileAsync(fileName);
                        }
                        catch (Exception e) { }

                    }

                    if (file != null)
                    {
                        var stream = await file.OpenAsync(FileAccessMode.Read);

                        var mediaSource = MediaSource.CreateFromStorageFile(file);
                        player.Source = mediaSource;
                    }
                }).AsTask();

            return player;
        }

        [ReactMethod]
        public void play(int key, ICallback callback)
        {
            Boolean callbackWasCalled = false;
            MediaPlayer player = null;
            Debug.WriteLine("play()");

            if (!playerPool.TryGetValue(key, out player))
            {
                Debug.WriteLine("Player is null");
                callback.Invoke(false);
                return;
            }

            if (player == null)
            {
                Debug.WriteLine("Player is null");
                callback.Invoke(false);
                return;
            }
            if (player.PlaybackSession.PlaybackState == MediaPlaybackState.Playing)
            {
                Debug.WriteLine("Already playing...");
                return;
            }

            player.MediaEnded +=
                delegate
                {
                    if (callbackWasCalled) return;
                    callbackWasCalled = true;
                    Debug.WriteLine("Media Ended");
                    callback.Invoke(true);
                };

            player.Play();
        }

        [ReactMethod]
        public void pause(int key, ICallback callback)
        {
            MediaPlayer player = null;

            if (!playerPool.TryGetValue(key, out player))
            {
                return;
            }
            if (player != null && player.PlaybackSession.PlaybackState == MediaPlaybackState.Playing)
            {
                player.Pause();
            }
            callback.Invoke();
        }

        [ReactMethod]
        public void stop(int key, ICallback callback)
        {
            MediaPlayer player = null;

            if (!playerPool.TryGetValue(key, out player))
            {
                return;
            }

            player.Pause();
            player.PlaybackSession.Position = new TimeSpan(0);
            callback.Invoke();
        }

        [ReactMethod]
        public void release(int key)
        {
            MediaPlayer player = null;

            if (!playerPool.TryGetValue(key, out player))
            {
                return;
            }

            player.Source = null;
            playerPool.Remove(key);
        }

        [ReactMethod]
        public void setVolume(int key, float volume)
        {
            MediaPlayer player = null;

            if (!playerPool.TryGetValue(key, out player))
            {
                return;
            }

            player.Volume = volume;


        }

        [ReactMethod]
        public void setLooping(int key, bool looping)
        {
            MediaPlayer player = null;

            if (!playerPool.TryGetValue(key, out player))
            {
                return;
            }

            player.IsLoopingEnabled = looping;

        }

        [ReactMethod]
        public void setCurrentTime(int key, float seconds)
        {
            MediaPlayer player = null;

            if (!playerPool.TryGetValue(key, out player))
            {
                return;
            }

            player.PlaybackSession.Position = new TimeSpan(0, 0, 0, 0, (int)seconds * 1000);

        }

        [ReactMethod]
        public void getCurrentTime(int key, ICallback callback)
        {
            MediaPlayer player = null;

            if (!playerPool.TryGetValue(key, out player))
            {
                return;
            }
            callback.Invoke(player.PlaybackSession.Position.Milliseconds * .001, player.PlaybackSession.PlaybackState == MediaPlaybackState.Playing);
        }

        [ReactMethod]
        public void enable(bool enabled)
        {
            // no op
        }
    }
}

File: windows/RNSound/RNSound/RNSound.cpp
#include "pch.h"
#include "RNSound.h"

using namespace winrt::Windows::ApplicationModel;
using namespace winrt::Windows::Media::Core;
using namespace winrt::Windows::Media::Playback;
using namespace winrt::Windows::Storage;
using namespace winrt::RNSound;

void RNSound::Prepare(const std::string& fileName, int key, JSValueObject options,
    std::function<void(JSValue, JSValue)> callback) noexcept {
    bool enableSMTCIntegration = true;
    if (options["enableSMTCIntegration"] != nullptr) {
        enableSMTCIntegration = options["enableSMTCIntegration"].AsBoolean();
    }

    MediaPlayer player{};
    loadFile(player, fileName);

    player.CommandManager().IsEnabled(enableSMTCIntegration);

    if (player == nullptr) {
        JSValueObject e{};
        e["code"] = -1;
        e["message"] = "resource not found";
        callback(std::move(e), nullptr);
        return;
    }

    player.AutoPlay(false);
    player.MediaOpened(
        [callback](winrt::Windows::Media::Playback::MediaPlayer sender,
               const winrt::Windows::Foundation::IInspectable &) {
            JSValueObject props{};
            props["duration"] = sender.PlaybackSession().NaturalDuration().count() / 
                winrt::impl::filetime_period::den;
            callback(nullptr, std::move(props));
        });

    playerPool[key] = player;
}

winrt::Windows::Foundation::IAsyncAction RNSound::loadFile(MediaPlayer player,
    const std::string &fileName) {
    if (fileName.rfind("http", 0) == 0)
    {
        winrt::Windows::Foundation::Uri uri{ winrt::to_hstring(fileName) };
        auto mediaSource = MediaSource::CreateFromUri(uri);
        player.Source(mediaSource);
    }
    else
    {
        StorageFile file{nullptr};
        StorageFolder LocalFolder = ApplicationData::Current().LocalFolder();
        StorageFolder InstallationFolder =
            winrt::Windows::ApplicationModel::Package::Current()
            .InstalledLocation();

        try {
            file = co_await winrt::Windows::ApplicationModel::Package::Current()
                .InstalledLocation().GetFileAsync(winrt::to_hstring(fileName));
        }
        catch (winrt::hresult_error const& ex) {
            (void)ex;
        }

        if (file == nullptr) {
            try {
                auto filePath = std::string("Assets\\") + fileName;
                file = co_await winrt::Windows::ApplicationModel::Package::Current()
                    .InstalledLocation()
                    .GetFileAsync(winrt::to_hstring(filePath));
            }
            catch (winrt::hresult_error const& ex) {
                (void)ex;
            }
        }

        if (file == nullptr) {
            try {
                file =
                    co_await LocalFolder.GetFileAsync(winrt::to_hstring(fileName));
            }
            catch (winrt::hresult_error const& ex) {
                (void)ex;
            }
        }

        if (file != nullptr) {
            auto stream = co_await file.OpenAsync(FileAccessMode::Read);

            auto mediaSource = MediaSource::CreateFromStorageFile(file);
            player.Source(mediaSource);
        }
    }
}

void RNSound::Play(int key,
                   std::function<void(bool)> callback) noexcept {
    bool callbackWasCalled = false;

    OutputDebugStringA("play()");

    if (playerPool.find(key) == playerPool.end()) {
        OutputDebugStringA("Player is null");
        callback(false);
        return;
    }

    auto player = playerPool[key];
    if (player == nullptr) {
        OutputDebugStringA("Player is null");
        callback(false);
        return;
    }

    if (player.PlaybackSession().PlaybackState() == MediaPlaybackState::Playing) {
        OutputDebugStringA("Already playing...");
        return;
    }

    player.MediaEnded(
        [callback, &callbackWasCalled](winrt::Windows::Media::Playback::MediaPlayer sender,
            const winrt::Windows::Foundation::IInspectable &) {
            if (callbackWasCalled)
                return;
            callbackWasCalled = true;
            OutputDebugStringA("Media Ended");
            callback(true);
        });

    player.Play();
}

void RNSound::Pause(int key, std::function<void()> callback) noexcept {
    if (playerPool.find(key) == playerPool.end()) {
        return;
    }

    auto player = playerPool[key];
    if (player != nullptr && player.PlaybackSession().PlaybackState() ==
                                 MediaPlaybackState::Playing) {
        player.Pause();
    }

    callback();
}

void RNSound::Stop(int key, std::function<void()> callback) noexcept {

    if (playerPool.find(key) == playerPool.end()) {
        return;
    }

    auto player = playerPool[key];
    if (player != nullptr) {
        player.Pause();
        player.PlaybackSession().Position(winrt::Windows::Foundation::TimeSpan(0));
        callback();
    }
}

void RNSound::Release(int key) noexcept {
    if (playerPool.find(key) == playerPool.end()) {
        return;
    }

    auto player = playerPool[key];
    if (player != nullptr) {
        player.Source(nullptr);
    }

    playerPool.erase(key);
}

void RNSound::SetVolume(int key, float volume) noexcept {
    if (playerPool.find(key) == playerPool.end()) {
        return;
    }

    auto player = playerPool[key];
    if (player != nullptr) {
        player.Volume(volume);
    }
}

void RNSound::SetLooping(int key, bool looping) noexcept {
    if (playerPool.find(key) == playerPool.end()) {
        return;
    }

    auto player = playerPool[key];
    if (player != nullptr) {
        player.IsLoopingEnabled(looping);
    }
}

void RNSound::SetCurrentTime(int key, float seconds) noexcept {
    if (playerPool.find(key) == playerPool.end()) {
        return;
    }

    auto player = playerPool[key];
    if (player != nullptr) {
        player.PlaybackSession().Position(winrt::Windows::Foundation::TimeSpan((int)seconds * 1000));
    }
}

void RNSound::GetCurrentTime(int key,
                             std::function<void(float, bool)> callback) noexcept {
    if (playerPool.find(key) == playerPool.end()) {
        return;
    }

    auto player = playerPool[key];
    if (player != nullptr) {
        callback(player.PlaybackSession().Position().count() * .001f,
                        player.PlaybackSession().PlaybackState() ==
                            MediaPlaybackState::Playing);
    }
}

void RNSound::Enable(bool enable) noexcept {
    // no op
    (void)enable;
}